README.md

<!-- README.md is generated from README.Rmd with devtools::build_readme(). Please edit that file -->

# genecovr

<!-- badges: start -->

[![R build
status](https://github.com/NBISweden/genecovr/workflows/R-CMD-check/badge.svg)](https://github.com/NBISweden/genecovr/actions)
<!-- badges: end -->

`genecovr` is an `R` package that provides plotting functions that
summarize gene transcript to genome alignments. The main purpose is to
assess the effect of polishing and scaffolding operations has on the
quality of a genome assembly. The gene transcript set is a large
sequence set consisting of assembled transcripts from RNA-seq data
generated in relation to a genome assembly project. Therefore,
`genecovr` serves as a complement to software such as
[BUSCO](https://busco.ezlab.org/), which evaluates genome assembly
quality using a smaller set of well-defined single-copy orthologs.

## Installation

You can install the released version of genecovr from [NBIS
GitHub](https://github.com/nbis) with:

``` r
# If necessary, uncomment to install devtools
# install.packages("devtools")
devtools::install_github("NBISweden/genecovr")
```

## Usage

### genecovr script quick start

There is a helper script for generating basic plots located in
PACKAGE_DIR/bin/genecovr. Create a data input csv-delimited file with
columns

1.  data label
2.  mapping file (supported formats: psl)
3.  assembly file (fasta or fasta index)
4.  transcript file (fasta or fasta index)

Columns 3 and 4 can be set to missing value (NA) in which case sequence
sizes will be inferred from the alignment files. Then run the script to
generate plots:

``` shell
PACKAGE_DIR/bin/genecovr indata.csv
```

#### Example

There are example files located in PACKAGE_DIR/inst/extdata consisting
of two psl alignment files containing gmap alignments and fasta indices
for the transcript sequences and two for different assembly versions:

- nonpolished.fai - fasta index for raw assembly
- polished.fai - fasta index for polished assembly
- transcripts.fai - fasta index for transcript sequences
- transcripts2nonpolished.psl - gmap alignments, transcripts to raw
  assembly
- transcripts2polished.psl - gmap alignments, transcripts to polished
  assembly

Using these files and the labels `non` and `pol` for the different
assemblies, a `genecovr` input file (called e.g., `assemblies.csv`)
would look as follows:

    nonpol,transcripts2nonpolished.psl,nonpolished.fai,transcripts.fai
    pol,transcripts2polished.psl,polished.fai,transcripts.fai

and the command to run would be:

    genecovr assemblies.csv

#### genecovr options

To list genecovr script options, type ’genecovr -h\`:

    usage: genecovr [-h] [-v] [-p number]
                                 [-d OUTPUT_DIRECTORY] [--height HEIGHT]
                                 [--width WIDTH]
                                 csvfile

    positional arguments:
      csvfile               csv-delimited file with columns
                                1. data label
                                2. mapping file (supported formats: psl)
                                3. assembly file (fasta or fasta index)
                                4. transcript file (fasta or fasta index)

    optional arguments:
      -h, --help            show this help message and exit
      -v, --verbose         print extra output
      -p number, --cpus number
                            number of cpus [default 1]
      -d OUTPUT_DIRECTORY, --output-directory OUTPUT_DIRECTORY
                            output directory
      --height HEIGHT       figure height in inches [default 6.0]
      --width WIDTH         figure width in inches [default 6.0]

### R package vignette

Alternatively, import the library in an R script and use the package
functions. See [Get started](articles/genecovr.html) or run
`vignette("genecovr")` for a minimum working example.


NEWS.md
<!-- markdownlint-disable MD025 -->

# genecovr 0.1.1

- update README
- add Empirical studies section

# genecovr 0.1.0

- add pkgdown site

# genecovr 0.0.0.9013

- fix factor level ordering for geneBodyCoverage plot
- save geneBodyCoverage as tsv

# genecovr 0.0.0.9012

- adjust factor levels for number of inserts (#4)
- summarize number of inserts by transcript (#5)

# genecovr 0.0.0.9011

- fix order of factors

# genecovr 0.0.0.9010

- remove duplicate entries in psl input

# genecovr 0.0.0.9009

- add plot of transcript length distributions conditioned on number of
  mapped contigs
- fix y axis for histogram plots
- increase point size of some plots
- summarizeGeneBodyCoverage and countSubjectsByCoverage now accept
  DataFrame inputs, obviating the need to rerun geneBodyCoverage
  multiple times in genecovr script

# genecovr 0.0.0.9008

- Remove characters trailing first space in fasta headers

# genecovr 0.0.0.9007

- Fix conversion of DNAStringSet to Seqinfo
- Make sure geneBodyCoverage table has nmax levels

# genecovr 0.0.0.9006

- add depthOfCoverage function and analysis to vignette and script
- reduceHitCoverage is deprecated
- improve some docs
- add wrapper for saving plots
- add tests mainly for alignmentpairs and test setup

<!-- markdownlint-enable MD025 -->


tests/testthat.R
library(testthat)
library(genecovr)

test_check("genecovr")


tests/testthat/setup.R
##################################################
## Small queries and subjects
##
## Cover as many corner cases as possible with simple examples
##################################################

##############################
## Multiple queries to one subject
##############################
ap1 <- AlignmentPairs(
    query = GenomicRanges::GRanges(
                               ranges = IRanges::IRanges(
                                                     start = c(1, 1, 90),
                                                     end = c(95, 98, 98)),
                               seqnames = c("t1.1", "t1.2", "t1.1"),
                               seqinfo = GenomeInfoDb::Seqinfo(seqnames = c("t1.1", "t1.2"),
                                                               seqlengths = c(100, 110))
                           ),
    subject = GenomicRanges::GRanges(
                                 ranges = IRanges::IRanges(
                                                       start = c(300, 300, 100),
                                                       end = c(400, 400, 200)),
                                 seqnames = c("ctg1.1", "ctg1.1", "ctg1.2"),
                                 seqinfo = GenomeInfoDb::Seqinfo(seqnames = c("ctg1.1", "ctg1.2"),
                                                                 seqlengths = c(500, 250))),
    matches = c(90, 90, 4))


##############################
## Multiple subjects to one query
##
## depthOfCoverage / breadthOfCoverage ~ 2 (i.e. duplication)
##############################
ap2 <- AlignmentPairs(
    query = GenomicRanges::GRanges(
                               ranges = IRanges::IRanges(
                                                     start = c(1, 3),
                                                     end = c(95, 98)),
                               seqnames = c("t2.1", "t2.1"),
                               seqinfo = GenomeInfoDb::Seqinfo(seqnames = c("t2.1"),
                                                               seqlengths = c(100))
                           ),
    subject = GenomicRanges::GRanges(
                                 ranges = IRanges::IRanges(
                                                       start = c(300, 300),
                                                       end = c(400, 400)),
                                 seqnames = c("ctg2.1", "ctg2.2"),
                                 seqinfo = GenomeInfoDb::Seqinfo(seqnames = c("ctg2.1", "ctg2.2"),
                                                                 seqlengths = c(500, 600))),
    matches = c(80, 90))


##############################
## Query split on different subjects
##
## depthOfCoverage / breadthOfCoverage ~ 1 (i.e. split)
##############################
ap3 <- AlignmentPairs(
    query = GenomicRanges::GRanges(
                               ranges = IRanges::IRanges(
                                                     start = c(1, 51),
                                                     end = c(54, 98)),
                               seqnames = c("t3.1", "t3.1"),
                               seqinfo = GenomeInfoDb::Seqinfo(seqnames = c("t3.1"),
                                                               seqlengths = c(100))
                           ),
    subject = GenomicRanges::GRanges(
                                 ranges = IRanges::IRanges(
                                                       start = c(100, 100),
                                                       end = c(160, 160)),
                                 seqnames = c("ctg3.1", "ctg3.2"),
                                 seqinfo = GenomeInfoDb::Seqinfo(seqnames = c("ctg3.1", "ctg3.2"),
                                                                 seqlengths = c(200, 200))),
    matches = c(50, 45))



##############################
## Slightly larger AlignmentPairs
##############################
ap4 <- AlignmentPairs(
    query = GenomicRanges::GRanges(
                               ranges = IRanges::IRanges(
                                                     start = c(1, 1, 90, 1, 115, 1, 190),
                                                     end = c(95, 98, 98, 120, 190, 170, 200)),
                               seqnames = c("t4.1", "t4.2", "t4.1", "t4.3", "t4.3", "t4.3", "t4.3"),
                               seqinfo = GenomeInfoDb::Seqinfo(seqnames = c("t4.1", "t4.2", "t4.3"),
                                                               seqlengths = c(100, 110, 200))
                           ),
    subject = GenomicRanges::GRanges(
                                 ranges = IRanges::IRanges(
                                                       start = c(300, 300, 100, 100, 220, 200, 300),
                                                       end = c(400, 400, 200, 225, 310, 380, 310)),
                                 seqnames = c("ctg4.1", "ctg4.1", "ctg4.2", "ctg4.3", "ctg4.3", "ctg4.4", "ctg4.5"),
                                 seqinfo = GenomeInfoDb::Seqinfo(seqnames = c("ctg4.1", "ctg4.2", "ctg4.3", "ctg4.4", "ctg4.5"),
                                                                 seqlengths = c(500, 250, 400, 500, 400)),
                                 NumInsert = c(),
                                 BaseInsert = c(),
                             ),
    matches = c(90, 90, 4, 100, 60, 160, 10))


##############################
## AlignmentPairs with numinserts
##############################
ap5 <- AlignmentPairs(
    query = GenomicRanges::GRanges(
                               ranges = IRanges::IRanges(
                                                     start = c(1, 81, 10, 90, 1, 50, 150),
                                                     end = c(80, 100, 100, 110, 100, 170, 200)),
                               seqnames = c("t5.1", "t5.1", "t5.2", "t5.2", "t5.3", "t5.3", "t5.3"),
                               seqinfo = GenomeInfoDb::Seqinfo(seqnames = c("t5.1", "t5.2", "t5.3"),
                                                               seqlengths = c(100, 110, 200)),
                               NumInsert = c(2, 0, 1, 4, 2, 3, 0),
                               BaseInsert = c(2, 0, 2, 6, 3, 3, 0),
                           ),
    subject = GenomicRanges::GRanges(
                                 ranges = IRanges::IRanges(
                                                       start = c(300, 401, 100, 200, 221, 200, 300),
                                                       end = c(380, 420, 190, 220, 320, 320, 350)),
                                 seqnames = c("ctg5.1", "ctg5.1", "ctg5.2", "ctg5.3", "ctg5.3", "ctg5.4", "ctg5.5"),
                                 seqinfo = GenomeInfoDb::Seqinfo(seqnames = c("ctg5.1", "ctg5.2", "ctg5.3", "ctg5.4", "ctg5.5"),
                                                                 seqlengths = c(500, 250, 400, 500, 400)),
                                 NumInsert = c(5, 1, 1, 5, 4, 3, 0),
                                 BaseInsert = c(6, 2, 1, 6, 4, 5, 0),
                             ),
    matches = c(78, 20, 90, 5, 97, 116, 50)
)




##############################
## AligmnentPairsList
##############################
apl <- AlignmentPairsList(list(ap1, ap2, ap3))

##################################################
## System files
##################################################
assembly_fai_fn <- list(
    nonpol = system.file("extdata", "nonpolished.fai", package = "genecovr"),
    pol = system.file("extdata", "polished.fai", package = "genecovr")
)
transcripts_fai_fn <- list(
    nonpol = system.file("extdata", "transcripts.fai", package = "genecovr"),
    pol = system.file("extdata", "transcripts.fai", package = "genecovr")
)

psl_fn <- list(
    nonpol = system.file("extdata", "transcripts2nonpolished.psl",
                         package = "genecovr"),
    pol = system.file("extdata", "transcripts2polished.psl",
                      package = "genecovr")
)


withr::defer(rm(assembly_fai_fn, transcripts_fai_fn, psl_fn))


tests/testthat/test-psl.R
## FIXME: add tests


R/AllClasses.R
setClassUnion("characterOrMissing", c("character", "missing", "logical"))
setClassUnion("factorOrcharacter", c("factor", "character"))
setClassUnion("integerOrMissing", c("integer", "missing", "logical"))


.valid.AlignmentPairs <- function(object) {
    if (length(object)) {
        if (!("query" %in% colnames(elementMetadata(object))))
            return(paste0("AlignmentPairs must have a 'query' column"))
        if (!("subject" %in% colnames(elementMetadata(object))))
            return(paste0("AlignmentPairs must have a 'subject' column"))
    }
}

##' Representation of an alignment pair
##'
##' @description Pairs subclass
##'
##'
##' @details The AlignmentPairs class extends the
##'     \code{\link[S4Vectors]{Pairs-class}}
##'
##' @export
##' @rdname AlignmentPairs-class
##'
##' @import S4Vectors
##'
##' @seealso \code{\link[S4Vectors]{Pairs-class}}
##'
setClass("AlignmentPairs",
         contains = c("Pairs"),
         validity = .valid.AlignmentPairs)


##' @importFrom methods validObject callNextMethod
setMethod("initialize", "AlignmentPairs", function(.Object, ...) {
    .Object <- callNextMethod()
    validObject(.Object)
    .Object
})


##' List of AlignmentPairs instances
##'
##' @description Subclass of S4Vectors SimpleList, where each entry is
##'     an AlignmentPairs object
##'
##' @export
##' @rdname AlignmentPairsList-class
##'
##' @import S4Vectors
##'
setClass("AlignmentPairsList",
         contains = "SimpleList",
         prototype = prototype(elementType = "AlignmentPairs")
         )


R/AllGenerics.R
##' @section Constructor:
##'
##' \code{AlignmentPairs(query, subject, ...)}: Constructs an AlignmentPairs
##' object by aligning the vectors \code{query} and \code{subject}.
##'
##' @param query vectorized object representing query
##' @param subject vectorized object representing subject
##' @param ... Arguments to pass to constructor
##'
##' @return AlignmentPairs object
##'
##' @export
##' @rdname AlignmentPairs-class
##'
setGeneric("AlignmentPairs", signature = c("query", "subject"),
           function(query, subject, ...)
    standardGeneric("AlignmentPairs"))


##' Retrieve score from object
##'
##' @description For any object with a score slot retrieve the score
##'     object.
##'
##' @param x object representing an alignment
##' @param ... Additional arguments
##'
##' @return numeric
##'
##' @export
##' @rdname score-methods
##'
##'
setGeneric("score", function(x, ...) standardGeneric("score"))

##' Retrieve query from object
##'
##' @description For any object with a query slot retrieve the query
##'     object.
##'
##' @param x object representing an alignment
##' @param ... Additional arguments
##'
##' @return query object
##'
##' @export
##' @rdname query-methods
##'
setGeneric("query", function(x, ...) standardGeneric("query"))

##' Set the query of an object
##'
##' @description For any object with a query slot set the
##'     query.
##'
##' @param x object representing an alignment
##' @param value object representing a query
##'
##' @return object representing an alignment with an updated query slot
##'
##' @export
##' @rdname query-methods
##'
setGeneric("query<-", function(x, value) standardGeneric("query<-"))

##' Retrieve subject from object
##'
##' @description For any object with a subject slot retrieve the
##'     subject object.
##'
##' @param x object representing an alignment
##' @param ... Additional arguments
##'
##' @return object representing a subject
##'
##' @export
##' @rdname sbjct-methods
##'
setGeneric("sbjct", function(x, ...) standardGeneric("sbjct"))

##' Set the subject of an object
##'
##' @description For any object with a subject slot set the
##'     subject.
##'
##' @param x object representing an alignment
##' @param value object representing a subject
##'
##' @return object representing an alignment with an updated subject
##'     slot
##'
##' @export
##' @rdname sbjct-methods
##'
setGeneric("sbjct<-", function(x, value) standardGeneric("sbjct<-"))

##' Get divergence from an alignment
##'
##' @description Get divergence from an alignment.
##'
##' @param x Object representation of a sequence alignment
##' @param ... Additional argument list
##'
##' @return numeric
##'
##' @export
##' @rdname divergence-methods
##'
##'
setGeneric("divergence", function(x, ...) standardGeneric("divergence"))

##' Get deletions from an alignment.
##'
##' @description Get deletions from an alignment.
##'
##' @param x Object representation of a sequence alignment
##' @param ... Additional argument list
##'
##' @return numeric
##'
##' @export
##' @docType methods
##' @rdname deletions-methods
##'
setGeneric("deletions", function(x, ...) standardGeneric("deletions"))

##' Get insertions from an alignment
##'
##' @description Get insertions from an alignment.
##'
##' @param x Object representation of a sequence alignment
##' @param ... Additional argument list
##'
##' @return numeric
##'
##' @export
##' @docType methods
##' @rdname insertions-methods
##'
setGeneric("insertions", function(x, ...) standardGeneric("insertions"))

##' matches
##'
##' @description Get matches from an alignment.
##'
##' @param x Object representation of a sequence alignment
##' @param ... Additional argument list
##'
##' @return numeric
##'
##' @export
##' @rdname matches-methods
##'
setGeneric("matches", function(x, ...) standardGeneric("matches"))

##' Get mismatches from an alignment
##'
##' @description Get mismatches from an alignment.
##'
##' @param x Object representation of a sequence alignment
##' @param ... Additional argument list
##'
##' @return numeric
##' @export
##' @rdname matches-methods
##'
setGeneric("mismatches", function(x, ...) standardGeneric("mismatches"))

##' Get matches that are parts of repeats from an alignment
##'
##' @description Get matches that are parts of repeats from an alignment.
##'
##' @param x Object representation of a sequence alignment
##' @param ... Additional argument list
##'
##' @return numeric
##'
##' @export
##' @rdname matches-methods
##'
setGeneric("repmatches", function(x, ...) standardGeneric("repmatches"))


##' @section Constructor:
##'
##' \code{AlignmentPairsList(obj, ...)}: Constructs an
##' AlignmentPairsList object from the input
##'
##' @param obj list
##' @param ... ellipsis
##'
##' @export
##' @rdname AlignmentPairsList-class
##'
##' @return AlignmentPairsList object
##'
setGeneric("AlignmentPairsList", signature = c("obj"),
           function(obj, ...)
    standardGeneric("AlignmentPairsList"))

##' Get gene body coverage
##'
##' \code{geneBodyCoverage(obj, ...)}: calculate gene body coverage from input
##'
##' @param obj alignment object
##' @param ... additional parameters
##'
##'
##' @export
##' @rdname geneBodyCoverage
##'
setGeneric("geneBodyCoverage", signature = c("obj"),
           function(obj, ...)
    standardGeneric("geneBodyCoverage"))


R/methods-sequence-data.R
##' methods-sequence-data
##'
##' @description Methods for dealing with sequence data
##'
##'
##' @name sequence data methods
##'
NULL


##' readFastaIndex
##'
##' @description Read a fasta index file
##'
##' @param fai fasta index file
##' @param ... additional parameters to pass to GenomeInfoDb::Seqinfo
##'
##' @return Seqinfo object
##'
##' @import GenomeInfoDb
##'
##' @export
##' @rdname readFastaIndex
##'
##' @examples
##' fai <- system.file("extdata", "polished.fai",
##'        package="genecovr")
##' sinfo <- readFastaIndex(fai)
##'
readFastaIndex <- function(fai, ...) {
    data <- read.table(fai, header = FALSE,
                       col.names = c("NAME", "LENGTH", "OFFSET",
                                     "LINEBASES", "LINEWIDTH"),
                       as.is = TRUE)
    GenomeInfoDb::Seqinfo(data$NAME, data$LENGTH, ...)
}


R/methods-AlignmentPairs-class.R
##' AlignmentPairs
##'
##' @description Constructor for AlignmentPairs
##'
##' @export
##' @rdname AlignmentPairs-class
##'
##' @importFrom GenomicRanges GRanges
##' @importFrom S4Vectors DataFrame
##'
setMethod("AlignmentPairs", signature = c("GRanges", "GRanges"),
          definition = function(query, subject, ...) {
    if (!missing(...)) {
        elementMetadata <- DataFrame(...)
    } else {
        elementMetadata <- new("DFrame", nrows = length(query))
    }
    elementMetadata$query <- query
    elementMetadata$subject <- subject
    new("AlignmentPairs",
        first = seq_along(query),
        second = seq_along(subject),
        elementMetadata = elementMetadata)
})

##############################
## Getters
##############################

##'
##' @export
##' @rdname query-methods
##'
setMethod("query", "AlignmentPairs", function(x) mcols(x)$query)

##'
##' @export
##' @rdname query-methods
##'
##' @importFrom methods validObject
##'
setMethod("query<-",
          signature = c("AlignmentPairs", "GRanges"),
          function(x, value) {
    mcols(x)$query <- value
    validObject(x)
    x
})

##'
##' @export
##' @rdname sbjct-methods
##'
setMethod("sbjct", "AlignmentPairs", function(x) mcols(x)$subject)

##'
##' @export
##' @rdname sbjct-methods
##'
##' @importFrom methods validObject
##'
setMethod("sbjct<-",
          signature = c("AlignmentPairs", "GRanges"),
          function(x, value) {
    mcols(x)$subject <- value
    validObject(x)
    x
})


##'
##' @export
##' @rdname matches-methods
##'
setMethod("matches", "AlignmentPairs",
          function(x) {
    retval <- NULL
    if ("matches" %in% names(values(x)))
        retval <- values(x)$matches
    retval
})

##'
##' @export
##' @rdname matches-methods
##'
setMethod("mismatches", "AlignmentPairs",
          function(x) {
    retval <- NULL
    if ("mismatches" %in% names(values(x)))
        retval <- values(x)$mismatches
    else if ("misMatches" %in% names(values(x)))
        retval <- values(x)$misMatches
    retval
})

##'
##' @export
##' @rdname matches-methods
##'
setMethod("repmatches", "AlignmentPairs",
          function(x) {
    retval <- NULL
    if ("repmatches" %in% names(values(x)))
        retval <- values(x)$repmatches
    else if ("repMatches" %in% names(values(x)))
        retval <- values(x)$repMatches
    retval
})



##' reduceHitCoverage
##'
##' @description Reduce hit coverage for a seqname
##'
##' @details Given an AlignmentPairs object, reduce overlapping query
##'     regions. For instance, if one hit spans coordinates 100-400
##'     and another 200-500, we merge to 100-500. The function
##'     calculates how many overlaps that are merged, the rationale
##'     being that if there are >1 overlaps for a seqname, the
##'     sequence is associated with multiple distinct regions in the
##'     subject sequence. Thus, the unit of interest here is the
##'     query, and the function seeks to identify reduced disjoint
##'     regions of a query sequence that map to a subject.
##'
##'     The function returns a GRanges object with reduced regions,
##'     i.e. overlaps are merged. Information about matches and
##'     mismatches is currently dropped as it usually is not possible
##'     to infer where mismatches occur. Instead, the data column
##'     `coverage` simply holds the ratio of the width of the region
##'     to the transcript length. Summing up coverages from disjoint
##'     regions then gives total coverage of the transcript.
##'
##'     The cutoff is used to filter regions based on the ratio of
##'     matches to the width of the region.
##'
##'     Since the association between query and subject regions is
##'     removed, the return value is a GRanges object consisting of
##'     the reduced query ranges with a revmap and coverage attribute.
##'
##' @rdname reduceHitCoverage
##' @export
##'
##' @importFrom IRanges CharacterList reduce width
##' @importFrom S4Vectors subset
##'
##' @param x AlignmentPairs object
##' @param min.match filter out hits with fraction of matching bases
##'     less than min.match
##'
##' @return reduced and filtered GRanges object
##'
##' @examples
##' ranges <- IRanges::IRanges(
##'           start=c(100, 200, 700),
##'           end=c(400, 500, 1000)
##' )
##' qry <- GenomicRanges::GRanges(
##'           ranges=ranges,
##'           seqnames=c("t1"),
##'           seqinfo=GenomeInfoDb::Seqinfo(seqnames=c("t1"),
##'                                         seqlengths=c(1050))
##' )
##' ranges <- IRanges::IRanges(
##'           start=c(1000, 1000, 1000),
##'           end=c(1300, 1300, 1300)
##' )
##' sbj <- GenomicRanges::GRanges(ranges=ranges, seqnames=c("c1", "c2", "c1"))
##' x <- AlignmentPairs(query=qry, subject=sbj, matches=c(300, 290, 280))
##' \dontrun{gr <- reduceHitCoverage(x, 0.1)}
##'
reduceHitCoverage <- function(x, min.match=0.9) {
    .Deprecated()
    stopifnot(inherits(x, "AlignmentPairs"))
    message("Reducing hits for ", summary(x), ", min.match ", min.match)
    y <- subset(x, matches(x) / width(query(x)) > min.match)
    q <- reduce(query(y), ignore.strand = TRUE,
                with.revmap = TRUE)
    q
}

##' breadthOfCoverage
##'
##' @description Calculate breadth of coverage by sequence id
##'
##' @param x GRangesList object
##'
##' @rdname breadthOfCoverage
##' @export
##'
##' @importFrom GenomicRanges GRangesList
##' @importFrom IRanges width reduce
##'
##' @return integer representing the sum of range widths
##'
## see https://stackoverflow.com/questions/39401376/aggregate-bins-in-large-granges-efficiently
breadthOfCoverage <- function(x) {
    stopifnot(inherits(x, "GRangesList"))
    sum(width(reduce(x)))
}

##' depthOfCoverage
##'
##' @description Calculate depth of coverage by sequence id
##'
##' @param x GRangesList object
##'
##' @rdname depthOfCoverage
##' @export
##'
##' @importFrom GenomicRanges GRangesList
##' @importFrom IRanges width
##'
##' @return integer representing the sum of range widths
##'
## see https://stackoverflow.com/questions/39401376/aggregate-bins-in-large-granges-efficiently
depthOfCoverage <- function(x) {
    stopifnot(inherits(x, "GRangesList"))
    sum(width(x))
}


##' geneBodyCoverage
##'
##' @description Calculate geneBodyCoverage of query sequences
##'
##' @details Given an AlignmentPairs object, calculate gene body
##'     coverages of query sequences.
##'     regions. For instance, if one hit spans coordinates 100-400
##'     and another 200-500, we merge to 100-500. The function
##'     calculates how many overlaps that are merged, the rationale
##'     being that if there are >1 overlaps for a seqname, the
##'     sequence is associated with multiple distinct regions in the
##'     subject sequence. Thus, the unit of interest here is the
##'     query, and the function seeks to identify reduced disjoint
##'     regions of a query sequence that map to a subject.
##'
##'     The function returns a GRanges object with reduced regions,
##'     i.e. overlaps are merged. Information about matches and
##'     mismatches is currently dropped as it usually is not possible
##'     to infer where mismatches occur. Instead, the data column
##'     `coverage` simply holds the ratio of the width of the region
##'     to the transcript length. Summing up coverages from disjoint
##'     regions then gives total coverage of the transcript.
##'
##'     The cutoff is used to filter regions based on the ratio of
##'     matches to the width of the region.
##'
##'     Since the association between query and subject regions is
##'     removed, the return value is a GRanges object consisting of
##'     the reduced query ranges with a revmap and coverage attribute.
##'
##'
##' @rdname geneBodyCoverage
##' @export
##'
##' @param obj AlignmentPairs object
##' @param min.match filter out hits with fraction matching bases less
##'     than min.match
##'
##' @importFrom IRanges IntegerList
##' @importFrom S4Vectors DataFrame
##'
##' @return reduced and filtered DataFrame object where each row is a
##'     transcript. See @details for information on seqnames and
##'     seqlengths columns are collected from the corresponding
##'     seqinfo object. breadthOfCoverage is the total width of all
##'     reduced regions of the transcript and corresponds to how much
##'     of a transcript is covered. Dividing breadthOfCoverage by the
##'     total transcript length gives the coverage fraction.
##'     depthOfCoverage sums all ranges and is an estimate of how many
##'     times each range is present. Dividing the depthOfCoverage by
##'     breadthOfCoverage indicates the multiplicity of the query. For
##'     queries that map multiple subjects, a value close to 1
##'     indicates the query has been split in several subjects,
##'     whereas a higher value indicates sequence duplication at the
##'     subject level.
##'
##'     The revmap column maps the output ranges to the input ranges
##'     as lists of numerical ids. These ids can be used to retrieve
##'     the corresponding AlignmentPairs ranges providing a link to
##'     the subjects. The hitCoverage lists the width of each reduced
##'     hit, and hitStart and hitEnd provide the transcript
##'     coordinates of these hits.
##'
setMethod("geneBodyCoverage", signature = c("AlignmentPairs"),
          definition = function(obj, min.match=0.9) {
    message("Calculating gene body coverage for ",
            summary(obj), ", min.match ", min.match)
    x <- subset(obj, matches(obj) / width(query(obj)) > min.match)
    y <- reduce(query(x), with.revmap=TRUE, ignore.strand=TRUE)
    grl <- split(query(x), seqnames(query(x)))
    revmapToList <- function(x) unique(unlist(x$revmap))
    data <- DataFrame(
        seqnames = seqnames(seqinfo(query(x))),
        seqlengths = seqlengths(seqinfo(query(x))),
        breadthOfCoverage = breadthOfCoverage(grl),
        depthOfCoverage = depthOfCoverage(grl),
        revmap = IRanges::IntegerList(lapply(split(y, seqnames(y)), revmapToList)),
        hitCoverage = width(reduce(grl)),
        hitStart = start(grl),
        hitEnd = end(grl)
    )
    data$coverage <- data$breadthOfCoverage / data$seqlengths
    data$revmap.count <- unlist(lapply(data$revmap, length))
    numinsert <- function(i) sum(insertions(query(x)[i]))
    data$NumInsert <- unlist(lapply(data$revmap, numinsert))
    baseinsert <- function(i) sum(insertions(query(x)[i], as.bases=TRUE))
    data$BaseInsert <- unlist(lapply(data$revmap, baseinsert))
    data$n.subjects <- unlist(lapply(data$revmap, function(j) {length(seqnames(sbjct(x[j])))}))
    metadata(data) <- list(analysis="geneBodyCoverage", min.match=min.match)
    data
})


##' summarizeGeneBodyCoverage
##'
##' @rdname summarizeGeneBodyCoverage
##' @export
##'
##' @importFrom S4Vectors DataFrame metadata
##'
##' @param x AlignmentPairs or DataFrame object
##' @param min.coverage coverage cutoffs to apply to transcripts
##' @param min.match filter out hits with fraction matching bases less
##'     than min.match
##'
summarizeGeneBodyCoverage <- function(x, min.coverage=seq(0, 1, 0.05),
                                      min.match=0.9) {
    stopifnot(inherits(x, "AlignmentPairs") | inherits(x, "DataFrame"))
    message("Summarizing gene body coverage for ",
            summary(x), ", ", length(min.coverage), " coverage cutoffs")
    if (inherits(x, "DataFrame")) {
        stopifnot(metadata(x)$analysis == "geneBodyCoverage")
        min.match <- metadata(x)$min.match
    } else
        x <- geneBodyCoverage(x, min.match)
    DataFrame(
        count = rev(cumsum(rev(apply(
            table(x$coverage, cut(x$coverage, min.coverage)), 2, sum)))),
        total = nrow(x),
        min.coverage = min.coverage[2:length(min.coverage)],
        min.match = min.match
    )
}


## FIXME: the method needs a rewrite!
##' countSubjectsByCoverage
##'
##' @rdname countSubjectsByCoverage
##' @export
##'
##' @importFrom S4Vectors DataFrame metadata
##'
##' @param x AlignmentPairs or DataFrame object
##' @param min.coverage coverage cutoffs to apply to transcripts
##' @param min.match filter out hits with fraction matching bases less
##'     than min.match
##' @param nmax maximum number of subjects to count
##'
countSubjectsByCoverage <- function(x, min.coverage=seq(0.25, 1, 0.25),
                                    min.match=0.9, nmax=5) {
    stopifnot(inherits(x, "AlignmentPairs") | inherits(x, "DataFrame"))
    message("Counting number of subjects by coverage for ",
            summary(x), ", ", length(min.coverage), " coverage cutoffs")
    if (inherits(x, "DataFrame")) {
        stopifnot(metadata(x)$analysis == "geneBodyCoverage")
        min.match <- metadata(x)$min.match
    } else
        x <- geneBodyCoverage(x, min.match)
    tab <- table(factor(
        x$n.subjects,
        levels = sort(unique(c(0, 1:max(x$n.subjects, nmax))))),
        cut(x$coverage, min.coverage))[, (length(min.coverage) - 1):1]
    z <- t(as.table(apply(tab, 1, cumsum)))
    z[1, ] <- nrow(x) - apply(z[2:nrow(z), ], 2, sum)
    z.df <- DataFrame(z)
    colnames(z.df) <- c("n.subjects", "min.coverage", "Freq")
    z.df$min.match <- min.match
    if (nmax < max(as.numeric(levels(z.df$n.subjects))))
        nmax <- max(as.numeric(levels(z.df$n.subjects)))
    z.df$n.subjects <- factor(z.df$n.subjects, levels = c(0, nmax:1))
    levels(z.df$n.subjects)[1] <- "filtered"
    levels(z.df$min.coverage) <- rev(min.coverage[1:(length(min.coverage) - 1)])
    z.df$min.coverage <- factor(
        z.df$min.coverage,
        levels = min.coverage[1:(length(min.coverage) - 1)])
    z.df
}
